_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+ - Hold a pointer to an object with automatic storage duration (an 'on-stack' object) after the end of the call in which it was created.
+ - Hold a temporarily delegated pointer beyond a single call.
+ - Modify an object passed via immutable reference.
+ - Modify any object reachable from an object that is passed as a deeply immutable reference.
+ - Tamper with (or access) an object passed via opaque reference.
+
+Some of these are baseline guarantees from CHERI, some are built atop those with additional hardware extensions, and others are built atop them with some language-agnostic software.
+Let's go through them one at a time.
+
+Running code cannot access objects unless passed pointers to them.
+This is a basic property of CHERI but also the absolute minimum foundation for *any* form of memory safety.
+If malicious or buggy code can materialise arbitrary pointers out of thin air and use them, it can bypass any of the other rules.
+In any CHERI system, all instructions that accesses memory (loads, stores, jumps, and so on) must take a capability as one of its operands.
+A capability (in general) is an unforgeable token of authority that allows you to do some action.
+A CHERI capability is a machine word that is interpreted by the CPU and which can be used as a hardware type that *represents the language notion of a pointer*.
+CHERI capabilities were designed to allow everything that correct C does with pointers, which is a superset of what most safer languages permit.
+
+That capability is protected both in registers and memory by a tag bit.
+The tag bit is an attestation from the hardware that there is a valid chain of provenance from the root capabilities that authorise everything down via a sequence of subsetting operations that end up with the capability that you hold.
+On CHERIoT, for example, if you have a capability to a heap object, this is telling you that the loader restricted a read-write capability to cover the address space that's used for the heap, then the allocator subset it to give you a pointer to an object.
+The CPU doesn't know any of the bits about the software model, of course, it just ensures that you didn't create a fake valid pointer.
+The software model defines what valid paths exist between the initial boot state and normal execution with you holding a pointer to a heap object, the hardware guarantees that some such path must have existed for you to hold that pointer.
+
+The two guarantees about immutability (code may not modify an object passed via immutable reference or any object any object reachable from an object that is passed as a deeply immutable reference) are simple properties once you have unforgeable CHERI capabilities.
+These are both enforced via *permissions*.
+Each capability carries a set of permissions that defines the set of things that it can be used for.
+Removing store permission means that it can't be used with store instructions (and so gives a read-only pointer).
```

Comment:
When deriving a new capability from an existing one you can only keep or remove permission; if you are only passed a read-only pointer you have no way to turn it back into a read-write pointer.